.github/workflows/check.yml

```diff
@@ -1,44 +1,50 @@
 name: Check
 
-on:
-  push:
-    branches:
-      - main
-  pull_request:
-    branches:
-      - main
+on: [ push, pull_request ]
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}-${{ github.event_name }}
+  cancel-in-progress: true
 
 jobs:
-  check:
+  rustfmt:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v6
+      with:
+        submodules: "recursive"
+    - name: Install Rust
+      uses: dtolnay/rust-toolchain@master
+      with:
+        toolchain: nightly
+        components: rustfmt
+    - name: Check formatting
+      run: cargo fmt --all -- --check
+  clippy:
     runs-on: ubuntu-latest
     strategy:
       fail-fast: false
       matrix:
         include:
-          - arch: riscv64
-            target: riscv64gc-unknown-none-elf
-          - arch: loongarch64
-            target: loongarch64-unknown-none-softfloat
-          - arch: aarch64
-            target: aarch64-unknown-none-softfloat
-          - arch: x86_64
-            target: x86_64-unknown-none
-
+        - arch: riscv64
+          target: riscv64gc-unknown-none-elf
+          args: ""
+        - arch: loongarch64
+          target: loongarch64-unknown-none-softfloat
+          args: "--skip vf2"
+        - arch: aarch64
+          target: aarch64-unknown-none-softfloat
+          args: "--skip vf2"
+        - arch: x86_64
+          target: x86_64-unknown-none
+          args: "--skip vf2"
     steps:
-      - uses: actions/checkout@v4
-        with:
-          submodules: "recursive"
-
-      - uses: Swatinem/rust-cache@v2
-        with:
-          key: ${{ matrix.arch }}
-
-      - uses: arceos-org/setup-musl@v1
-        with:
-          arch: ${{ matrix.arch }}
-
-      - name: Clippy
-        run: cargo clippy --target ${{ matrix.target }} -F qemu -- -D warnings
+    - uses: actions/checkout@v6
+      with:
+        submodules: "recursive"
+    - uses: arceos-org/setup-musl@v1
+      with:
+        arch: ${{ matrix.arch }}
+    - uses: taiki-e/install-action@cargo-hack
+    - name: Run Clippy
+      run: cargo hack --target ${{ matrix.target }} --each-feature --keep-going --workspace ${{matrix.args}} clippy -- -D warnings
```

Comment:
`cargo hack` is a great tool. But to keep this PR simple, please don't modify the clippy part in this PR yet.